File: backend/app/schemas/analytics.py
"""Pydantic models and TypedDicts for analytics API endpoints."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class CVEEnrichmentResult(BaseModel):
    """Result of CVE enrichment data processing from process_cve_enrichments()."""

    max_epss: float | None = None
    max_percentile: float | None = None
    max_risk: float | None = None
    has_kev: bool = False
    kev_count: int = 0
    kev_ransomware_use: bool = False
    kev_due_date: str | None = None
    exploit_maturity: str = "unknown"
    days_until_due: int | None = None


class SeverityBreakdown(BaseModel):
    """Breakdown of findings by severity level."""

    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    negligible: int = 0
    info: int = 0
    unknown: int = 0


class DependencyUsage(BaseModel):
    """Usage statistics for a dependency across projects."""

    name: str
    type: str
    versions: list[str]
    project_count: int
    total_occurrences: int
    has_vulnerabilities: bool
    vulnerability_count: int


class DependencyTreeNode(BaseModel):
    """A dependency and the ids of its direct children; the client expands children lazily."""

    id: str
    name: str
    version: str
    purl: str
    type: str
    direct: bool
    direct_inferred: bool = False
    has_findings: bool
    findings_count: int
    findings_severity: SeverityBreakdown | None = None
    child_ids: list[str] = []
    source_type: str | None = None
    source_target: str | None = None
    layer_digest: str | None = None
    locations: list[str] = []


class DependencyGraph(BaseModel):
    """Flat unique dependency nodes + the ids to render at top level; the client nests lazily."""

    # Every node is reachable from roots (direct deps, unresolved-parent deps, one entry per
    # otherwise-disconnected component), so the graph is rendered whole without server nesting.
    nodes: list[DependencyTreeNode] = []
    roots: list[str] = []


class ImpactAnalysisResult(BaseModel):
    """Result of impact analysis for a component."""

    component: str
    version: str
    affected_projects: int
    total_findings: int
    findings_by_severity: SeverityBreakdown
    recommended_version: str | None = None
    fix_impact_score: float
    affected_project_names: list[str]
    max_epss_score: float | None = None
    epss_percentile: float | None = None
    has_kev: bool = False
    kev_count: int = 0
    kev_ransomware_use: bool = False
    kev_due_date: str | None = None  # Earliest CISA remediation deadline
    days_until_due: int | None = None  # negative when overdue
    exploit_maturity: str = "unknown"
    max_risk_score: float | None = None
    days_known: int | None = None
    has_fix: bool = False
    fix_versions: list[str] = []
    priority_reasons: list[str] = []


class VulnerabilityHotspot(BaseModel):
    """A vulnerability hotspot - component with many findings."""

    component: str
    version: str
    type: str
    finding_count: int
    severity_breakdown: SeverityBreakdown
    affected_projects: list[str]
    first_seen: str
    max_epss_score: float | None = None
    epss_percentile: float | None = None
    has_kev: bool = False
    kev_count: int = 0
    kev_ransomware_use: bool = False
    kev_due_date: str | None = None  # Earliest CISA remediation deadline
    days_until_due: int | None = None  # negative when overdue
    exploit_maturity: str = "unknown"
    max_risk_score: float | None = None
    days_known: int | None = None
    has_fix: bool = False
    fix_versions: list[str] = []
    top_cves: list[str] = []
    priority_reasons: list[str] = []


class DependencyTypeStats(BaseModel):
    """Statistics for a dependency type."""

    type: str
    count: int
    percentage: float


class AnalyticsSummary(BaseModel):
    """Summary of analytics across all accessible projects."""

    total_dependencies: int
    total_vulnerabilities: int
    unique_packages: int
    dependency_types: list[DependencyTypeStats]
    severity_distribution: SeverityBreakdown


class DependencyMetadata(BaseModel):
    """Aggregated metadata for a dependency across all projects."""

    name: str
    version: str
    type: str
    purl: str | None = None

    description: str | None = None
    author: str | None = None
    publisher: str | None = None
    homepage: str | None = None
    repository_url: str | None = None
    download_url: str | None = None
    group: str | None = None

    license: str | None = None
    license_url: str | None = None
    license_category: str | None = None
    license_risks: list[str] = []
    license_obligations: list[str] = []

    deps_dev: dict[str, Any] | None = None

    project_count: int = 0
    affected_projects: list[dict[str, Any]] = []  # [{id, name, direct}]
    total_vulnerability_count: int = 0
    total_finding_count: int = 0

    enrichment_sources: list[str] = []


class VulnerabilitySearchResult(BaseModel):
    """Result of a vulnerability/CVE search."""

    vulnerability_id: str  # e.g. CVE-2021-44228, GHSA-xxx
    aliases: list[str] = []

    severity: str
    cvss_score: float | None = None
    epss_score: float | None = None
    epss_percentile: float | None = None

    in_kev: bool = False
    kev_ransomware: bool = False
    kev_due_date: str | None = None

    component: str
    version: str
    component_type: str | None = None
    purl: str | None = None

    project_id: str
    project_name: str
    scan_id: str | None = None

    finding_id: str
    finding_type: str
    description: str | None = None
    fixed_version: str | None = None

    waived: bool = False
    waiver_reason: str | None = None


class VulnerabilitySearchResponse(BaseModel):
    """Paginated response for vulnerability search."""

    items: list[VulnerabilitySearchResult]
    total: int
    page: int
    size: int


class DependencySearchResult(BaseModel):
    """Result of a dependency search."""

    project_id: str
    project_name: str
    package: str
    version: str
    type: str
    license: str | None = None
    license_url: str | None = None
    direct: bool = False
    purl: str | None = None
    source_type: str | None = None
    source_target: str | None = None
    layer_digest: str | None = None
    found_by: str | None = None
    locations: list[str] = []
    cpes: list[str] = []
    description: str | None = None
    author: str | None = None
    publisher: str | None = None
    group: str | None = None
    homepage: str | None = None
    repository_url: str | None = None
    download_url: str | None = None
    hashes: dict[str, Any] = {}
    properties: dict[str, Any] = {}


class DependencySearchResponse(BaseModel):
    """Paginated response for dependency search."""

    items: list[DependencySearchResult]
    total: int
    page: int
    size: int


class RecommendationResponse(BaseModel):
    """Response model for a single recommendation."""

    type: str
    priority: str
    title: str
    description: str
    impact: dict[str, Any]
    affected_components: list[str]
    action: dict[str, Any]
    effort: str


class RecommendationsResponse(BaseModel):
    """Response model for recommendations endpoint."""

    project_id: str
    project_name: str
    scan_id: str
    total_findings: int
    total_vulnerabilities: int
    recommendations: list[RecommendationResponse]
    summary: dict[str, Any]


# --- Update Frequency Analysis ---


class DependencyUpdateEvent(BaseModel):
    """A single dependency version change between two consecutive scans."""

    package_name: str
    package_type: str
    purl: str | None = None
    old_version: str
    new_version: str
    update_type: str  # "patch" | "minor" | "major" | "unknown" | "downgrade"
    scan_date: str  # ISO timestamp of the scan where the update was detected
    previous_scan_date: str
    days_between_scans: int
    was_outdated: bool  # was flagged as outdated in the previous scan


class ScanTimelineEntry(BaseModel):
    """Per-scan summary for timeline chart data."""

    scan_id: str
    date: str
    updates_count: int  # excludes downgrades
    outdated_count: int
    patch: int
    minor: int
    major: int
    unknown: int = 0
    downgrades: int = 0


class SlowPackage(BaseModel):
    """A package that remains outdated across multiple scans."""

    name: str
    type: str
    current_version: str | None = None
    latest_version: str | None = None
    scans_outdated: int  # number of scans where this package was flagged as outdated


class UpdateFrequencyMetrics(BaseModel):
    """Aggregated update frequency metrics for a single project."""

    project_id: str
    project_name: str
    # Branch the timeline covers; None only when the project has no completed scans.
    branch: str | None = None
    scan_count: int
    time_range_days: float
    first_scan_date: str
    last_scan_date: str

    # Downgrades/rollbacks are tracked separately and never counted as updates.
    total_updates: int
    updates_per_scan: float
    updates_per_month: float  # extrapolated from time_range_days

    patch_updates: int
    minor_updates: int
    major_updates: int
    unknown_updates: int
    downgrade_updates: int = 0
    granularity_ratio: dict[str, float]  # {"patch": 0.6, "minor": 0.3, "major": 0.1}

    avg_days_between_scans: float

    # Coverage — share of ever-outdated packages that were resolved.
    # None means nothing was ever outdated (N/A, distinct from 0%).
    total_outdated_detected: int
    outdated_resolved: int
    update_coverage_pct: float | None = None

    trend_direction: str  # "improving" | "stable" | "deteriorating" | "unknown"
    trend_detail: str

    # Upstream release cadence (independent of scan frequency).
    # All four are None when no release-history data is available.
    upstream_releases_last_12m_median: float | None = None
    upstream_days_between_releases_median: float | None = None
    upstream_days_since_latest_release_median: float | None = None
    adoption_latency_days_median: float | None = None  # release-to-first-scan lag

    # Dominant dep ecosystem ("pypi"/"npm"/...); "mixed" if none >=70%; None if empty.
    dominant_ecosystem: str | None = None

    scan_timeline: list[ScanTimelineEntry]
    slowest_packages: list[SlowPackage]
    recent_updates: list[DependencyUpdateEvent]


class ProjectUpdateSummary(BaseModel):
    """Lightweight update metrics for cross-project comparison."""

    project_id: str
    project_name: str
    team_name: str | None = None
    scan_count: int
    updates_per_month: float
    update_coverage_pct: float | None = None
    patch_ratio: float  # proportion of patch updates (0-1)
    trend_direction: str  # "improving" | "stable" | "deteriorating" | "unknown"
    total_outdated: int
    last_scan_date: str


class UpdateFrequencyComparison(BaseModel):
    """Cross-project comparison of update frequency metrics.

    Ranking: measured coverage first (descending), then projects where
    nothing was ever outdated (no measurable coverage). best/worst are
    drawn from measured projects only.
    """

    projects: list[ProjectUpdateSummary]
    team_avg_updates_per_month: float
    team_avg_coverage_pct: float | None = None  # None when no project has measurable coverage
    best_project: str | None = None
    worst_project: str | None = None
    skipped_projects: int = 0  # projects dropped for lacking >=2 completed scans


# Crypto analytics schemas


class HotspotEntry(BaseModel):
    """A single entry in a crypto hotspot report."""

    key: str = Field(..., description="Grouping key (e.g., 'RSA-1024')")
    grouping_dimension: str = Field(..., description="Dimension this entry groups by")
    asset_count: int = Field(..., ge=0)
    finding_count: int = Field(..., ge=0)
    severity_mix: dict[str, int] = Field(default_factory=dict)
    locations: list[str] = Field(default_factory=list)
    project_ids: list[str] = Field(default_factory=list)
    first_seen: datetime
    last_seen: datetime


class HotspotResponse(BaseModel):
    """Paginated hotspot response for a given scope."""

    scope: Literal["project", "team", "global", "user"]
    scope_id: str | None = None
    grouping_dimension: str
    items: list[HotspotEntry] = Field(default_factory=list)
    total: int = Field(..., ge=0)
    generated_at: datetime
    cache_hit: bool = False


class TrendPoint(BaseModel):
    """A single data point in a trend time-series."""

    timestamp: datetime
    metric: str
    value: float


class TrendSeries(BaseModel):
    """A full trend time-series for a metric within a scope."""

    scope: str
    scope_id: str | None = None
    metric: str
    bucket: Literal["day", "week", "month"]
    points: list[TrendPoint] = Field(default_factory=list)
    range_start: datetime
    range_end: datetime
    cache_hit: bool = False

File: frontend/src/types/analytics.ts
import { Finding } from './scan';
import { Severity, SeverityBreakdown } from './common';

export type { Severity, SeverityBreakdown };

export interface DashboardStats {
    total_projects: number;
    total_critical: number;
    total_high: number;
    avg_risk_score: number;
    top_risky_projects: {
        name: string;
        risk: number;
        id: string;
    }[];
}

export interface SearchResult {
  project_id: string;
  project_name: string;
  package: string;
  version: string;
  type: string;
  license?: string;
  direct?: boolean;
  direct_inferred?: boolean;
  purl?: string;
  source_type?: string;
  source_target?: string;
  layer_digest?: string;
  locations?: string[];
}

export interface DependencyUsage {
  name: string;
  type: string;
  versions: string[];
  project_count: number;
  total_occurrences: number;
  has_vulnerabilities: boolean;
  vulnerability_count: number;
}

export interface DependencyTreeNode {
  id: string;
  name: string;
  version: string;
  purl: string;
  type: string;
  direct: boolean;
  direct_inferred?: boolean;
  has_findings: boolean;
  findings_count: number;
  findings_severity?: SeverityBreakdown;
  source_type?: string;
  source_target?: string;
  layer_digest?: string;
  locations?: string[];
  children: DependencyTreeNode[];
}

export interface ImpactAnalysisResult {
  component: string;
  version: string;
  affected_projects: number;
  total_findings: number;
  findings_by_severity: SeverityBreakdown;
  recommended_version?: string;
  fix_impact_score: number;
  affected_project_names: string[];
  max_epss_score?: number;
  epss_percentile?: number;
  has_kev?: boolean;
  kev_count?: number;
  kev_ransomware_use?: boolean;
  kev_due_date?: string;
  days_until_due?: number; 
  exploit_maturity?: string;
  max_risk_score?: number;
  days_known?: number;
  has_fix?: boolean;
  fix_versions?: string[];
  priority_reasons?: string[];
}

export interface VulnerabilityHotspot {
  component: string;
  version: string;
  type: string;
  finding_count: number;
  severity_breakdown: SeverityBreakdown;
  affected_projects: string[];
  first_seen: string;
  max_epss_score?: number;
  epss_percentile?: number;
  has_kev?: boolean;
  kev_count?: number;
  kev_ransomware_use?: boolean;
  kev_due_date?: string;
  days_until_due?: number;
  exploit_maturity?: string;
  max_risk_score?: number;
  days_known?: number;
  has_fix?: boolean;
  fix_versions?: string[];
  top_cves?: string[];
  priority_reasons?: string[]; 
}

export interface DependencyTypeStats {
  type: string;
  count: number;
  percentage: number;
}

export interface AnalyticsSummary {
  total_dependencies: number;
  total_vulnerabilities: number;
  unique_packages: number;
  dependency_types: DependencyTypeStats[];
  severity_distribution: SeverityBreakdown;
}

export interface AdvancedSearchResult {
  project_id: string;
  project_name: string;
  package: string;
  version: string;
  type: string;
  license?: string;
  license_url?: string;
  direct: boolean;
  direct_inferred?: boolean;
  purl?: string;
  source_type?: string;
  source_target?: string;
  layer_digest?: string;
  locations?: string[];
  cpes?: string[];
  description?: string;
  author?: string;
  publisher?: string;
  group?: string;
  homepage?: string;
  repository_url?: string;
  download_url?: string;
  hashes?: Record<string, string>;
  properties?: Record<string, string>;
  found_by?: string;
  license_category?: string; 
  licenses_detailed?: Array<{
    spdx_id: string;
    source: string;
    category?: string;
    explanation?: string;
  }>;
  license_risks?: string[];
  license_obligations?: string[];
  enrichment_sources?: string[];
  deps_dev?: {
    stars?: number;
    forks?: number;
    open_issues?: number;
    project_url?: string;
    project_description?: string;
    project_license?: string;
    dependents?: {
      total?: number;
      direct?: number;
      indirect?: number;
    };
    scorecard?: {
      overall_score?: number;
      date?: string;
      checks_count?: number;
    };
    links?: Record<string, string>;
    published_at?: string;
    is_deprecated?: boolean;
    known_advisories?: string[];
    has_attestations?: boolean;
    has_slsa_provenance?: boolean;
  };
}

export interface HotspotsQueryParams {
  skip?: number;
  limit?: number;
  sort_by?: 'finding_count' | 'component' | 'first_seen' | 'epss' | 'risk';
  sort_order?: 'asc' | 'desc';
}

export interface AdvancedSearchResponse {
  items: AdvancedSearchResult[];
  total: number;
  page: number;
  size: number;
}

export interface VulnerabilitySearchResult {
  vulnerability_id: string;
  aliases: string[];
  severity: Severity;
  cvss_score?: number;
  epss_score?: number;
  epss_percentile?: number;
  in_kev: boolean;
  kev_ransomware: boolean;
  kev_due_date?: string;
  component: string;
  version: string;
  component_type?: string;
  purl?: string;
  project_id: string;
  project_name: string;
  scan_id?: string;
  finding_id: string;
  finding_type: string;
  description?: string;
  fixed_version?: string;
  waived: boolean;
  waiver_reason?: string;
}

export interface VulnerabilitySearchResponse {
  items: VulnerabilitySearchResult[];
  total: number;
  page: number;
  size: number;
}

export interface VulnerabilitySearchOptions {
  severity?: string;
  in_kev?: boolean;
  has_fix?: boolean;
  finding_type?: string;
  project_ids?: string[];
  include_waived?: boolean;
  sort_by?: 'severity' | 'cvss' | 'epss' | 'component' | 'project_name';
  sort_order?: 'asc' | 'desc';
  skip?: number;
  limit?: number;
}

export type ComponentFinding = Finding & { project_id: string; project_name: string; scan_id?: string };

export interface DependencyMetadata {
  name: string;
  version: string;
  type: string;
  purl?: string;
  description?: string;
  author?: string;
  publisher?: string;
  homepage?: string;
  repository_url?: string;
  download_url?: string;
  group?: string;
  license?: string;
  license_url?: string;
  license_category?: string;
  license_risks?: string[];
  license_obligations?: string[];
  deps_dev?: {
    stars?: number;
    forks?: number;
    open_issues?: number;
    is_deprecated?: boolean;
    known_advisories?: string[];
    published_at?: string;
    project_description?: string;
    project_url?: string;
    dependents?: { total?: number };
    scorecard?: {
      overall_score?: number;
      checks_count?: number;
      date?: string;
    };
    links?: Record<string, string>;
  };
  project_count: number;
  affected_projects: Array<{ id: string; name: string; direct: boolean; direct_inferred?: boolean }>;
  total_vulnerability_count: number;
  total_finding_count: number;
  enrichment_sources?: string[];
}

export type RecommendationType = 
  | 'base_image_update'
  | 'direct_dependency_update'
  | 'transitive_fix_via_parent'
  | 'no_fix_available'
  | 'consider_waiver'
  | 'rotate_secrets'
  | 'remove_secrets'
  | 'fix_code_security'
  | 'fix_infrastructure'
  | 'license_compliance'
  | 'supply_chain_risk'
  | 'outdated_dependency'
  | 'version_fragmentation'
  | 'dev_in_production'
  | 'unmaintained_package'
  | 'recurring_vulnerability'
  | 'regression_detected'
  | 'deep_dependency_chain'
  | 'duplicate_functionality'
  | 'cross_project_pattern'
  | 'shared_vulnerability';

export type RecommendationPriority = 'critical' | 'high' | 'medium' | 'low';

export interface RecommendationImpact {
  critical: number;
  high: number;
  medium: number;
  low: number;
  total: number;
}

export interface CrossProjectCve {
  cve: string;
  total_affected: number;
  affected_projects?: string[];
}

export interface RecommendationAction {
  type: string;
  package?: string;
  current_version?: string;
  target_version?: string;
  current_image?: string;
  suggestion?: string;
  commands?: string[];
  cves?: Array<string | CrossProjectCve>;
  options?: string[];
  suggestions?: string[];
  file_path?: string;
  line_number?: number;
  secret_type?: string;
  files?: string[];
  rule_ids?: string[];
  license_type?: string;
  components?: string[];
  resource_type?: string;
  description?: string;
  packages?: Array<{
    name: string;
    current?: string;
    recommended_major?: number;
    reason?: string;
    versions?: string[];
    suggestion?: string;
    version_count?: number;
    project_count?: number;
  }>;
  new_critical_cves?: string[];
  delta?: number;
  deepest_chains?: Array<{
    package: string;
    depth: number;
    chain_preview?: string;
  }>;
  duplicates?: Array<{
    category: string;
    found: string[];
    suggestion: string;
  }>;
  affected_projects?: string[];
  total_affected?: number;
  priority_projects?: Array<{
    name: string;
    id: string;
    critical: number;
    high: number;
  }>;
}

export interface Recommendation {
  type: RecommendationType;
  priority: RecommendationPriority;
  title: string;
  description: string;
  impact: RecommendationImpact;
  affected_components: string[];
  affected_projects?: Array<{ id: string; name: string }>;
  action: RecommendationAction;
  effort: 'low' | 'medium' | 'high';
}

export interface RecommendationsSummary {
  base_image_updates: number;
  direct_updates: number;
  transitive_updates: number;
  no_fix: number;
  total_fixable_vulns: number;
  total_unfixable_vulns: number;
  secrets_to_rotate: number;
  sast_issues: number;
  iac_issues: number;
  license_issues: number;
  quality_issues: number;
  outdated_deps?: number;
  fragmentation_issues?: number;
  trend_alerts?: number;
  cross_project_issues?: number;
}

export interface RecommendationsResponse {
  project_id: string;
  project_name: string;
  scan_id: string;
  total_findings: number;
  total_vulnerabilities: number;
  recommendations: Recommendation[];
  summary: RecommendationsSummary;
}

export interface AdvancedSearchOptions {
    version?: string;
    type?: string;
    source_type?: string;
    has_vulnerabilities?: boolean;
    project_ids?: string[];
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
    skip?: number;
    limit?: number;
}
